src/Compilers/Core/Portable/DiagnosticAnalyzer/CachingSemanticModelProvider.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System.Collections.Concurrent;
+
+namespace Microsoft.CodeAnalysis.Diagnostics
+{
+    internal sealed class CachingSemanticModelProvider : SemanticModelProvider
+    {
+        private readonly ConcurrentDictionary<SyntaxTree, SemanticModel> _semanticModelsMap;
+
+        public CachingSemanticModelProvider()
+        {
+            _semanticModelsMap = new ConcurrentDictionary<SyntaxTree, SemanticModel>();
+        }
+
+        public override SemanticModel GetSemanticModel(SyntaxTree tree, Compilation compilation)
+        {
+            if (_semanticModelsMap.TryGetValue(tree, out var model) &&
+                model.Compilation == compilation)
```

Comment:
>model.Compilation == compilation [](start = 16, length = 32)

Should we simply use compilation as part of the key?

Reply:
That seems reasonable, let me make that change.